core/arch/interp.c

```diff
@@ -8007,3 +8252,157 @@ d_r_emulate(dcontext_t *dcontext, app_pc pc, priv_mcontext_t *mc)
     instr_free(dcontext, &instr);
     return next_pc;
 }
+
+#ifdef AARCH64
+/* Emit addtional code to fix up indirect trace exit for AArch64.
+ * For each indirect branch in trace we have the following code:
+ *      str x0, TLS_REG0_SLOT
+ *      mov x0, #trace_next_target
+ *      eor x0, x0, jump_target
+ *      cbnz x0, trace_exit (ibl_routine)
+ *      ldr x0, TLS_REG0_SLOT
+ * For the trace_exit (ibl_routine), it needs to conform to the
+ * protocol specified in emit_indirect_branch_lookup in
+ * aarch64/emit_utils.c.
+ * The ibl routine requires:
+ *     x2: contains indirect branch target
+ *     TLS_REG2_SLOT: contains app's x2
+ * Therefore we need to add addtional spill instructions
+ * before we actually jump to the ibl routine.
+ * We want the indirect hit path to have minimum instructions
+ * and also conform to the protocol of ibl routine
+ * Therefore we append the restore at the end of the trace
+ * after the backward jump to trace head.
+ * For example, the code will be fixed to:
+ *      eor x0, x0, jump_target
+ *      cbnz x0, trace_exit_label
+ *      ...
+ *      b trace_head
+ * trace_exit_label:
+ *      ldr x0, TLS_REG0_SLOT
+ *      str x2, TLS_REG2_SLOT
+ *      mov x2, jump_target
+ *      b ibl_routine
+ */
+int
+fixup_indirect_trace_exit(dcontext_t *dcontext, instrlist_t *trace)
+{
+    instr_t *instr, *prev, *branch;
+    instr_t *trace_exit_label;
+    app_pc target = 0;
+    app_pc ind_target = 0;
+    reg_id_t scratch;
+    reg_id_t jump_target_reg = DR_REG_NULL;
+    uint indirect_type = 0;
+    int added_size = 0;
+    trace_exit_label = NULL;
+    instr_t *delete_after = NULL;
+    /* We record the original trace end */
+    instr_t *trace_end = instrlist_last(trace);
+
+    LOG(THREAD, LOG_MONITOR, 4, "fixup the indirect trace exit\n");
+
+    for (instr = instrlist_first(trace); instr != trace_end;
+         instr = instr_get_next(instr)) {
+        if (instr_is_exit_cti(instr)) {
+            target = instr_get_branch_target_pc(instr);
+            /* Check for indirect exit. */
+            if (is_indirect_branch_lookup_routine(dcontext, (cache_pc)target)) {
```

Comment:
`fixup_indirect_trace_exit()` is invoked each time we want to extend the trace, and it is possible for the new (potentially multiple) bbs to have multiple indirect exits, isn't it? Are you trying to say that once we have scanned some parts of the trace, next time we only need to scan those newly added bbs rather than the entire trace?

Reply:
Well, we would need to have multiple copies of `trace_exit:` even if `fixup_indirect_trace_exit` was invoked separately to fix up each exit, right? Because each fixed up indirect branch would still use its own `trace_exit`.

I'm not convinced whether we need to iterate through the complete trace. I think the question to answer is: how many indirect exits do we fix up  each time `fixup_indirect_trace_exit` is invoked. Can it be >1? Does that happen when multiple bbs are added to the trace together?
If we fix up only one exit in each invocation of `fixup_indirect_trace_exit`, then it does seem better to avoid the loop and see if we can get to the ibl routine branch some other way.

Can you add a comment on the loop, clarifying its need.

Reply:
I think it is indeed possible for multiple exits to be fixed when we invoke `mixup_indirect_trace_exit`. Here is an example trace fragment. I will add comments for clarification.
```
TAG  0x0000ffffad0385d4
 +0    L4 @0x0000fffd691e42c0  f9000380   str    %x0 -> (%x28)[8byte]
 +4    m4 @0x0000fffd691c4588  d29aee80   movz   $0xd774 lsl $0x00 -> %x0
 +8    m4 @0x0000fffd691c4f28  f2b52200   movk   %x0 $0xa910 lsl $0x10 -> %x0
 +12   m4 @0x0000fffd691c3338  f2dfffe0   movk   %x0 $0xffff lsl $0x20 -> %x0
 +16   L4 @0x0000fffd691c5128  ca1e0000   eor    %x0 %x30 lsl $0x00 -> %x0
 +20   L4 @0x0000fffd691e4450  b5000000   cbnz   $0x0000ffff6911ed80 <shared_trace_ibl_ret> %x0
 +24   L4 @0x0000fffd691c65b0  f9400380   ldr    (%x28)[8byte] -> %x0
 +28   L1                      a8c27bfd   ldp    (%sp)[16byte] %sp $0x0000000000000020 -> %x29 %
x30 %sp
 +32   L4 @0x0000fffd691c36f0  f9000380   str    %x0 -> (%x28)[8byte]
 +36   m4 @0x0000fffd691c5970  d29af880   movz   $0xd7c4 lsl $0x00 -> %x0
 +40   m4 @0x0000fffd691c43e0  f2b52200   movk   %x0 $0xa910 lsl $0x10 -> %x0
 +44   m4 @0x0000fffd691c3420  f2dfffe0   movk   %x0 $0xffff lsl $0x20 -> %x0
 +48   L4 @0x0000fffd691e5668  ca1e0000   eor    %x0 %x30 lsl $0x00 -> %x0
 +52   L4 @0x0000fffd691c6158  b5000000   cbnz   $0x0000ffff6911ed80 <shared_trace_ibl_ret> %x0
 +56   L4 @0x0000fffd691c3488  f9400380   ldr    (%x28)[8byte] -> %x0
 +60   L0                      910063e0   add    %sp $0x0018 lsl $0x00 -> %x0
 +64   L0                      91004001   add    %x0 $0x0010 lsl $0x00 -> %x1
 +68   L0                      910063e0   add    %sp $0x0018 lsl $0x00 -> %x0
 +72   L0                      91006000   add    %x0 $0x0018 lsl $0x00 -> %x0
 +76   L0                      d29afb1e   movz   $0xd7d8 lsl $0x00 -> %x30
 +80   L0                      f2b5221e   movk   %x30 $0xa910 lsl $0x10 -> %x30
 +84   L0                      f2dffffe   movk   %x30 $0xffff lsl $0x20 -> %x30
 +88   L4 @0x0000fffd691c5b38  14000000   b      $0x0000ffffa910d748
END 0x0000ffffad0385d4
```